### DISCUSSION TITLE
Why pageParams is undefined on init?

### DISCUSSION THREAD

**Discussion by behzadam** [TanStack/query#3442] [General] [answered]
I'm not sure this is a bug report or not but the question is why pageParams initial item is undefined? I expect it to be 1 and not undefined. I'm preparing a simple example of using React Query and Graphql and the demo is [here](https://codesandbox.io/s/pedantic-shamir-43wys8).
Also, you can check the issue in CharacterList component:
`  useEffect(() => {
    console.log(data?.pageParams)
  }, [data])`

<img width="359" alt="image" src="https://user-images.githubusercontent.com/5009172/160146036-1d1df737-ee21-4430-b420-53fa2ae6d3c8.png">

**Comment by TkDodo** [comment, marked as answer]
`pageParam` is derived from the _previous_ page via `getNextPageParam`. Since the first page has no previous page, what else should it be?

That is why we assign a default value in the example:

```
useInfiniteQuery(
    'projects',
    async ({ pageParam = 0 }) => {
      const res = await axios.get('/api/projects?cursor=' + pageParam)
      return res.data
    },
    {
      getNextPageParam: lastPage => lastPage.nextId ?? undefined,
    }
  )
```

in that case, `undefined` gets coerced to `0`

**Reply by behzadam** [reply]
I got you but when I try to make default like this:
`getNextPageParam: lastPage => lastPage.nextId ??  1`
I get undefined yet. I think it is not a big problem and even a problem but in a case, I wanted to make scrollable pagination with pageParam and just for previous pages not next pages, and then the first page was undefined for me. 
Of course, I know I can solve that problem in many ways but I wanted to share it with you that I can make any default for pageParam.

**Reply by TkDodo** [reply]
the misconception is probably that `getNextPageParam` is _not_ called at all for the first page. So it doesn't matter what you return as "default" here. If we were to call `getNextPageParam`, we would have to call it with `lastPage: undefined`, so your code would need to handle that via `lastPage?.nextId ?? 1`.

However, we actually _rely_ on `getNextPageParam` to return `undefined` to _stop_ fetching when the query is invalidated. All pages are refetched in order. However, you might have 5 pages in your cache, but because of deletion, there are now only 4. After fetching the first page, we call `getNextPageParam`, and use whatever we get returned here for the second page. Maybe the pageSize has changed, we don't know, so many things could happen in the meantime on the server :). So in my example, once we get to page 4, we will get back `undefined` from `getNextPageParam`, so page 5 will no longer be refetched.

Hope that makes sense and explains why we cannot let you return a default from  `getNextPageParam`.

**Reply by behzadam** [reply]
Thank you @TkDodo. Recently, I read the document in two days and tried to make a demo and your answers were helpful for me.

**Reply by amirfrd** [reply]
@TkDodo 
so in nextjs when try to prefetchInfiniteQuery because of that undefined can't return from getServerSideProps

`Error: Error serializing `.dehydratedState.queries[0].state.data.pageParams[0]` returned from `getServerSideProps` in "/store/[id]".
Reason: `undefined` cannot be serialized as JSON. Please use `null` or omit this value.`

**Reply by TkDodo** [reply]
it's fixed in v5

**Reply by sbbu** [reply]
can you fix it in v4

**Reply by TkDodo** [reply]
no
